README.md
<h1 align="center">
ScienceWorld
</h1>

<p align="center">
<!-- Version badge using shields.io -->
  <a href="https://github.com/allenai/ScienceWorld/releases">
    <img src="https://img.shields.io/github/v/release/allenai/ScienceWorld">
  </a>
<!-- Link to tutorials badge using shields.io -->
  <a href="https://huggingface.co/spaces/MarcCote/ScienceWorld">
    <img src="https://img.shields.io/badge/🤗-Demo-yellow">
  </a>
<!-- Follow on twitter badge using shields.io -->
  <a href="https://sciworld.apps.allenai.org">
    <img src="https://img.shields.io/badge/Website-green">
  </a>
</p>

ScienceWorld is a text-based virtual environment centered around accomplishing tasks from the standardized elementary science curriculum.  This code accompanies the paper [ScienceWorld: Is your Textual Agent Smarter than a 5th grader?](https://arxiv.org/abs/2203.07540).

<h3 align="center"><img src="https://github.com/allenai/ScienceWorld/blob/main/media/scienceworld_environment.png" width="75%"/></h3>

### Demo and examples

You can try ScienceWorld yourself via our [HuggingFace Space](https://huggingface.co/spaces/MarcCote/ScienceWorld) or read some of the [playthrough transcripts](https://sciworld.apps.allenai.org/explore).

### Citation
```
@misc{scienceworld2022,
    title={ScienceWorld: Is your Agent Smarter than a 5th Grader?},
    author={Ruoyao Wang and Peter Jansen and Marc-Alexandre C{\^o}t{\'e} and Prithviraj Ammanabrolu},
    year={2022},
    eprint={2203.07540},
    archivePrefix={arXiv},
    primaryClass={cs.CL},
    url={https://arxiv.org/abs/2203.07540}
}
```

# Quickstart
**Before running:** You will have to have `Java 1.8+` installed on your system (shipped with most linux distributions).

Install with pip:
```bash
conda create --name scienceworld python=3.8
conda activate scienceworld
pip install scienceworld
```

Run an example random agent, on task 13 (classification: place a non-living thing in a box), for 5 episodes:
> python examples/random_agent.py --task-num=13 --num-episodes=5 --simplifications-preset easy

Run a user console where you can interact with the environment, on task 3 (change of state: melting):
> python examples/human.py --task-num=3 --num-episodes=5


# Web Server Demo

A web server demo is also available, that allows running a ScienceWorld user console that can be interacted with in a web browser.

<h3 align="center"><img src="https://github.com/allenai/ScienceWorld/blob/main/media/web_demo_screenshot.png" width="75%"/></h3>

To run the web server demo:
```bash
conda create --name scienceworld python=3.8
conda activate scienceworld
pip install scienceworld[webserver]
```

Run the web server:
> python examples/scienceworld-web-server-example.py

Point your web browser to:
`localhost:8080`


# ScienceWorld Design
ScienceWorld is written in Scala (2.12.9), and compiles using `sbt` into a JAR file that is run with Java.  For convenience, a Python API is provided (Python >= 3.7), which interfaces using the `py4j` package.

# Tasks
The tasks are listed in the table below along with their number of variations. Either the task ID or its name can be used to a task with `env.load()`.

| Task ID | Task Name                                | # Variations |
|-------|----------------------------------------------------|------|
|   1-1 |                                             boil |   30 |
|   1-2 |                                             melt |   30 |
|   1-3 |                                           freeze |   30 |
|   1-4 |                    change-the-state-of-matter-of |   30 |
|   2-1 |                                  use-thermometer |  540 |
|   2-2 |            measure-melting-point-known-substance |  436 |
|   2-3 |          measure-melting-point-unknown-substance |  300 |
|   3-1 |                                  power-component |   20 |
|   3-2 | power-component-renewable-vs-nonrenewable-energy |   20 |
|   3-3 |                                test-conductivity |  900 |
|   3-4 |          test-conductivity-of-unknown-substances |  600 |
|   4-1 |                                find-living-thing |  300 |
|   4-2 |                            find-non-living-thing |  300 |
|   4-3 |                                       find-plant |  300 |
|   4-4 |                                      find-animal |  300 |
|   5-1 |                                       grow-plant |  126 |
|   5-2 |                                       grow-fruit |  126 |
|   6-1 |                                    chemistry-mix |   32 |
|   6-2 |              chemistry-mix-paint-secondary-color |   36 |
|   6-3 |               chemistry-mix-paint-tertiary-color |   36 |
|   7-1 |                           lifespan-longest-lived |  125 |
|   7-2 |                          lifespan-shortest-lived |  125 |
|   7-3 |       lifespan-longest-lived-then-shortest-lived |  125 |
|   8-1 |                           identify-life-stages-1 |   14 |
|   8-2 |                           identify-life-stages-2 |   10 |
|   9-1 |                   inclined-plane-determine-angle |  168 |
|   9-2 |           inclined-plane-friction-named-surfaces | 1386 |
|   9-3 |         inclined-plane-friction-unnamed-surfaces |  162 |
|  10-1 |                   mendelian-genetics-known-plant |  120 |
|  10-2 |                 mendelian-genetics-unknown-plant |  480 |

# Baseline Agents
**DRRN:** https://github.com/cognitiveailab/drrn-scienceworld

**KG-A2C:** https://github.com/cognitiveailab/kga2c-scienceworld

**CALM:** https://github.com/cognitiveailab/calm-scienceworld

**Behavior Cloning and Decision Transformer:** https://github.com/cognitiveailab/t5-scienceworld


CHANGELOG.md
# Changelog

## [1.1.3] - 2023-01-19
 - Improving how task name are handled (backward compatible). @aphedges ([#46](https://github.com/allenai/ScienceWorld/pull/46))
 - Set minimum supported Python version to 3.7. @aphedges ([#46](https://github.com/allenai/ScienceWorld/pull/46))

## [1.1.2] - 2022-12-20
 - Parity release. Task IDs from the paper (e.g., "2-3") are now accepted and the preferred way to load task. ([#44](https://github.com/allenai/ScienceWorld/pull/44))
 - Adding script to list all supported tasks in ScienceWorld. ([#44](https://github.com/allenai/ScienceWorld/pull/44))
 - Add environment variable to enable debugging mode. ([#43](https://github.com/allenai/ScienceWorld/pull/43))

## [1.1.1] - 2022-12-09
 - Improve README (badges, Hugginface Space).
 - Fix stack overflow with connected components. ([#41](https://github.com/allenai/ScienceWorld/pull/41))
 - Better package version management. ([#42](https://github.com/allenai/ScienceWorld/pull/42))

## [1.1.0] - 2022-11-23
 - **Breaking** Use py4j launch_gateway to launch Java server. ([#30](https://github.com/allenai/ScienceWorld/pull/30), [#35](https://github.com/allenai/ScienceWorld/pull/35))
 - Improved JAR building process. ([#22](https://github.com/allenai/ScienceWorld/pull/22))
 - Use logging.info instead of print. ([#36](https://github.com/allenai/ScienceWorld/pull/36))
 - Add temp unit. ([#38](https://github.com/allenai/ScienceWorld/pull/38))
 - Improve server example. ([#39](https://github.com/allenai/ScienceWorld/pull/39))
 - Make ScienceWorld more deterministic. ([#15](https://github.com/allenai/ScienceWorld/pull/15))
 - Fix typos. ([#26](https://github.com/allenai/ScienceWorld/pull/26),  [#28](https://github.com/allenai/ScienceWorld/pull/28), [#32](https://github.com/allenai/ScienceWorld/pull/32))
 - Add links for baseline agents used in the paper. ([#27](https://github.com/allenai/ScienceWorld/pull/27))
 - Fix JSON escape characters. ([#23](https://github.com/allenai/ScienceWorld/pull/23))

## [1.0.2] - 2022-06-16

 - Fixed issues with JAR path. ([#12](https://github.com/allenai/ScienceWorld/pull/12))

## [1.0.1] - 2022-06-07

 - Add versioning.

## [1.0.0] - 2022-05-31

- Initial release.

[1.1.2]: (https://github.com/allenai/ScienceWorld/tree/1.1.2)
[1.1.1]: (https://github.com/allenai/ScienceWorld/tree/1.1.1)
[1.1.0]: (https://github.com/allenai/ScienceWorld/tree/1.1.0)
[1.0.2]: (https://github.com/allenai/ScienceWorld/tree/1.0.2)
[1.0.1]: (https://github.com/allenai/ScienceWorld/tree/1.0.1)
[1.0.0]: (https://github.com/allenai/ScienceWorld/tree/1.0.0)


examples.py
#
#   Examples
#
#   conda create --name scienceworld python=3.8
#   conda activate scienceworld
#   pip install py4j                                (for scala-python interface)
#   pip install -U pywebio                          (for web server)


from scienceworld import ScienceWorldEnv
import argparse
import random
import timeit
import time

#
#   Example random agent -- randomly picks an action at each step.
#
def randomModel(args):
    exitCommands = ["quit", "exit"]    

    taskIdx = args['task_num']
    simplificationStr = args['simplification_str']
    numEpisodes = args['num_episodes']      

    # Keep track of the agent's final scores
    finalScores = []

    # Initialize environment
    env = ScienceWorldEnv("", args['jar_path'], envStepLimit = args['env_step_limit'] , threadNum = 0)
    taskNames = env.getTaskNames()
    print("Task Names: " + str(taskNames))

    # Choose task
    taskName = taskNames[taskIdx]        # Just get first task        
    env.load(taskName, 0, "")            # Load the task, so we have access to some extra accessors e.g. getRandomVariationTrain() )
    maxVariations = env.getMaxVariations(taskName)    
    print("Starting Task " + str(taskIdx) + ": " + taskName)
    time.sleep(2)

    # Start running episodes
    for episodeIdx in range(0, numEpisodes):        
        # Pick a random task variation
        randVariationIdx = env.getRandomVariationTrain()
        env.load(taskName, randVariationIdx, simplificationStr)

        # Reset the environment
        initialObs, initialDict = env.reset()
        
        # Example accessors
        print("Possible actions: " + str(env.getPossibleActions()) )
        print("Possible objects: " + str(env.getPossibleObjects()) )
        templates, lut = env.getPossibleActionObjectCombinations()
        print("Possible action/object combinations: " + str(templates))
        print("Object IDX to Object Referent LUT: " + str(lut))
        print("Task Name: " + taskName)
        print("Task Variation: " + str(randVariationIdx) + " / " + str(maxVariations))
        print("Task Description: " + str(env.getTaskDescription()) )        
        print("look: " + str(env.look()) )
        print("inventory: " + str(env.inventory()) )
        print("taskdescription: " + str(env.taskdescription()) )
        

        score = 0.0
        isCompleted = False
        curIter = 0        

        # Run one episode until we reach a stopping condition (including exceeding the maximum steps)
        userInputStr = "look around"        # First action
        while (userInputStr not in exitCommands) and (isCompleted == False):
            print("----------------------------------------------------------------")
            print ("Step: " + str(curIter))

            # Send user input, get response
            observation, score, isCompleted, info = env.step(userInputStr)
            print("\n>>> " + observation)
            print("Score: " + str(score))
            print("isCompleted: " + str(isCompleted))

            # The environment will makke isCompleted `True` when a stop condition has happened, or the maximum number of steps is reached.
            if (isCompleted):
                break

            # Randomly select action        

            ## Any action (valid or not)
            #templates, lut = env.getPossibleActionObjectCombinations()            
            #print("Possible action/object combinations: " + str(templates))
            #print("Object IDX to Object Referent LUT: " + str(lut))
            #randomTemplate = random.choice( templates )        
            #print("Next random action: " + str(randomTemplate))
            #userInputStr = randomTemplate["action"]

            ## Only valid actions
            validActions = env.getValidActionObjectCombinationsWithTemplates()
            randomAction = random.choice( validActions )        
            print("Next random action: " + str(randomAction))
            userInputStr = randomAction["action"]

                        
            print(list(lut.keys())[-1])

            # Sanitize input
            userInputStr = userInputStr.lower().strip()
            print("Choosing random action: " + str(userInputStr))

            # Keep track of the number of commands sent to the environment in this episode
            curIter += 1

        print("Goal Progress:")
        print(env.getGoalProgressStr())
        time.sleep(1)

        # Episode finished -- Record the final score
        finalScores.append(score)
            
        # Report progress of model        
        print ("Final score: " + str(score))
        print ("isCompleted: " + str(isCompleted))        

        # Save history -- and when we reach maxPerFile, export them to file
        filenameOutPrefix = args['output_path_prefix'] + str(taskIdx)        
        env.storeRunHistory(episodeIdx, notes = {'text':'my notes here'} )
        env.saveRunHistoriesBufferIfFull(filenameOutPrefix, maxPerFile=args['max_episode_per_file'])

    # Episodes are finished -- manually save any last histories still in the buffer
    env.saveRunHistoriesBufferIfFull(filenameOutPrefix, maxPerFile=args['max_episode_per_file'], forceSave=True)

    # Show final episode scores to user: 
    avg = sum([x for x in finalScores if x >=0]) / len(finalScores)     # Clip negative scores to 0 for average calculation
    print ("")
    print ("---------------------------------------------------------------------")
    print (" Summary (Random Agent)")
    print (" Task " + str(taskIdx) + ": " + taskName)
    print (" Simplifications: " + str(simplificationStr))
    print ("---------------------------------------------------------------------")
    print (" Epsiode scores: " + str(finalScores))    
    print (" Average episode score: " + str(avg))
    print ("---------------------------------------------------------------------")
    print ("")

    print("Shutting down server...")    
    env.shutdown()

    print("Completed.")


# Example user input console, to play through a game. 
def userConsole(args):
    exitCommands = ["quit", "exit"]    

    taskIdx = args['task_num']
    simplificationStr = args['simplification_str']      

    # Initialize environment
    env = ScienceWorldEnv("", args['jar_path'], envStepLimit = args['env_step_limit'] , threadNum = 0)
    taskNames = env.getTaskNames()
    print("Task Names: " + str(taskNames))

    # Choose task
    taskName = taskNames[taskIdx]
    env.load(taskName, args['var_num'], simplificationStr)
    print("Starting Task " + str(taskIdx) + ": " + taskName)
    time.sleep(2)    

    # Reset the environment
    initialObs, initialDict = env.reset()
    
    
    #
    #   Examples of how to access much of the environment information that the API exposes. 
    #   (Many of these are similar to the Jericho API)
    #
    print("Task Names: " + str(taskNames))
    print("Possible actions: " + str(env.getPossibleActions()) )
    print("Possible objects: " + str(env.getPossibleObjects()) )
    templates, lut = env.getPossibleActionObjectCombinations()
    print("Possible action/object combinations: " + str(templates))
    #print("Object IDX to Object Referent LUT: " + str(lut))
    print("Vocabulary: " + str(env.getVocabulary()) )
    print("Possible actions (with IDs): " + str(env.getPossibleActionsWithIDs()))
    print("Possible object types: " + str(env.getObjectTypes()))    
    print("Object IDX to Object Referent LUT: " + str(lut))
    print("\n")
    print("Possible object referents LUT: " + str(env.getPossibleObjectReferentLUT()))
    print("\n")
    print("Valid action-object combinations: " + str(env.getValidActionObjectCombinations()))
    print("\n")
    print("Object_ids to type_ids: " + str(env.getAllObjectTypesLUTJSON()))
    print("\n")
    print("All objects, their ids, types, and referents: " + str(env.getAllObjectIdsTypesReferentsLUTJSON() ))
    print("\n")
    print("Valid action-object combinations (with templates): " + str(env.getValidActionObjectCombinationsWithTemplates() ))
    print("\n")    
    print("Object Type LUT: " + str(env.getPossibleObjectReferentTypesLUT()))
    print("Variations (train): " + str(env.getVariationsTrain() ))    

    print("")
    print("----------------------------------------------------------------------------------")
    print("")

    print("Task Name: " + taskName)
    print("Variation: " + str(args['var_num']) + " / " + str(env.getMaxVariations(taskName)))
    print("Task Description: " + str(env.getTaskDescription()) )        
    
    #
    #   Main user input loop
    #
    userInputStr = "look around"        # First action
    while (userInputStr not in exitCommands):
        if (userInputStr == "help"):
            print("Possible actions: ")
            for actionStr in env.getPossibleActions():
                print("\t" + str(actionStr))

        elif (userInputStr == "objects"):
            print("Possible objects (one referent listed per object): ")
            for actionStr in env.getPossibleObjects():
                print("\t" + str(actionStr))

        elif (userInputStr == "valid"):        
            print("Valid action-object combinations:")
            print(env.getValidActionObjectCombinationsWithTemplates())

        elif (userInputStr == 'goals'):
            print(env.getGoalProgressStr())
            
        else:
            # Send user input, get response
            observation, score, isCompleted, info = env.step(userInputStr)
            print("\n" + observation)
            print("Score: " + str(score))
            print("isCompleted: " + str(isCompleted))
            #print("info: " + str(info))

        print("'help' lists valid action templates, 'objects' lists valid objects, 'valid' lists valid action-object combinations (long!). ")
        print("'goals' lists progress on subgoals.")
        print("type 'exit' to quit.")

        # Get user input        
        userInputStr = input('> ')
        # Sanitize input
        userInputStr = userInputStr.lower().strip()

    
    # Display run history     
    runHistory = env.getRunHistory()    
    print("Run History:")
    print(runHistory)
    for item in runHistory:
        print(item)
        print("")

    # Display subgoal progress
    print(env.getGoalProgressStr())


    print("Shutting down server...")    
    env.shutdown()

    print("Completed.")


#
#   Parse command line arguments
#
def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--jar_path", type=str)
    parser.add_argument("--task_num", type=int, default=13)
    parser.add_argument("--var_num", type=int, default=0)    
    parser.add_argument("--env_step_limit", type=int, default=100)    
    parser.add_argument("--num_episodes", type=int, default=5)    
    parser.add_argument("--simplification_str", default="easy")
    parser.add_argument("--max_episode_per_file", type=int, default=1000)
    parser.add_argument("--mode", default="randomagent")
    parser.add_argument("--output_path_prefix", default="save-histories")

    args = parser.parse_args()
    params = vars(args)
    return params


#
#   Main
#
def main():        
    print("ScienceWorld 1.0 API Examples")    

    # Parse command line arguments
    args = parse_args()

    # Run a user console
    if (args['mode'] == 'userconsole'):
        userConsole(args)

    # Run a model that chooses random actions until successfully reaching the goal
    if (args['mode'] == 'randomagent'):
        randomModel(args)

    print("Exiting.")

if __name__ == "__main__":
    main()
